### PR TITLE
Make catch blocks aware of all assignments in the try blocks

### PR DESCRIPTION
**Pull request by Kroppeb**

### CODE REVIEW

**Review by SuperCoder7979 — CHANGES_REQUESTED** [review]
I have a few suggestions and lingering questions for this PR, but otherwise this is great!

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/DirectNode.java [hunk 1]
```diff
@@ -33,22 +33,16 @@ public static DirectNode forStat(DirectNodeType type, Statement statement) {
...
-    this.statement = statement;
-
-    this.id = block.id.toString();
-    this.block = block;
-  }
-
-  public static DirectNode forBlock(Statement statement) {
-    return new DirectNode(DirectNodeType.DIRECT, statement, (BasicBlockStatement)statement);
+  public boolean hasSuccessors(DirectEdgeType type) {
+    return this.successors.containsKey(type) && !this.successors.get(type).isEmpty();
```
[the remark this answers is not in the document]
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/DirectNode.java:37]: `getSuccessors` does allocate an empty list if there is no list yet yet.

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java [hunk 2]
```diff
@@ -56,6 +57,25 @@ public DirectGraph buildDirectGraph(RootStatement root) {
...
+      directNode.block = (BasicBlockStatement) stat;
+    }
+    return directNode;
+  }
+
+  private DirectNode createDirectNode(Statement stat, DirectNodeType type) {
+    DirectNode node = DirectNode.forStat(type, stat);
+    this.graph.nodes.addWithKey(node, node.id);
+
+    if (!this.tryNodesStack.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:72]: That's currently being handled on [line 223](https://github.com/QuiltMC/quiltflower/pull/138/files#diff-8d96a692186ded9f7f28301b5ab815289ed3c2f7abdae8ea622778b29b72a4c0R223)

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java [hunk 3]
```diff
@@ -293,13 +332,13 @@ class StatementStackEntry {
                 // is essentially
                 //
                 // for (inc; ; init)
-                DirectNode inc = DirectNode.forStat(DirectNodeType.INCREMENT, stat);
-                graph.nodes.putWithKey(inc, inc.id);
+                // TODO: that ordering does not make sense
```
[the remark this answers is not in the document]
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:335]: Not really, for `for (a : b)` you'd expect b to be the init statement.

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/DirectNode.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (SuperCoder7979) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/DirectNode.java:37]: Fair. I was thinking that most nodes would have regular edges so it'd be faster, but most nodes probably won't have exception edges

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (SuperCoder7979) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:335]: If you see DoStatement, you'll see that it's inverted there. DoStatement says in `for (a : b)` a is the init and b is the increment

File: src/org/jetbrains/java/decompiler/modules/decompiler/sforms/SFormsConstructor.java [hunk 4]
```diff
@@ -744,6 +785,54 @@ private SFormsFastMapDirect getFilteredOutMap(String nodeid, String predid, Dire
         }
 
         mapNew.intersection(mapNewTemp);
+//
+//        if (this.trackPhantomExitNodes && !mapTrueSource.isEmpty() && !mapNew.isEmpty()) { // FIXME: what for??
```
Comment (SuperCoder7979) [inline src/org/jetbrains/java/decompiler/modules/decompiler/sforms/SFormsConstructor.java:789]: Forgot to add this in the review, but you should add a TODO here saying that we need to implement this at a later date, as discussed

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (SuperCoder7979) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:72]: I see, nice. A comment saying that's what it does would be good

File: src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kroppeb) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:335]: Yeah, I know. But just like you said "it's inverted". IHMO I feel like they should be swapped. It just something that has been bugging me for a long time.

[the remark this answers is not in the document]
Comment (SuperCoder7979) [reply src/org/jetbrains/java/decompiler/modules/decompiler/flow/FlattenStatementsHelper.java:335]: That's fair, we should definitely fix it sometime

**Review by SuperCoder7979 — APPROVED** [review]
Looks good!

**Pull request merged**